### PR TITLE
Add GL provider to surface_factory_starboard

### PR DESCRIPTION
**Pull request by TyHolc**
b/371272304

### CODE REVIEW

**Review by niranjanyardi — APPROVED** [review]
lgtm from a build perspective.

I don't understand the egl part of the code - might be good to get someone else's approval.

File: ui/ozone/platform/starboard/BUILD.gn
Comment (andrewsavage1) [inline ui/ozone/platform/starboard/BUILD.gn:38]: @niranjanyardi you'll probably want to start thinking about if we want people to depend on //starboard or //starboard:starboard_group after your work on the hermetic toolchain is complete—I'll assume we'll need to do the latter

Comment (niranjanyardi) [reply ui/ozone/platform/starboard/BUILD.gn:38]: latter +1

**Review by kaidokert — CHANGES_REQUESTED** [review]
New code module - should have a a unit test.  ( channeling my inner @yell0wd0g )

You can't _functionally_ test a lot here, as there's no obvious way to mock EGL, but at least we should have test code calling into the function.

**Review by y4vor — APPROVED** [review]

File: ui/ozone/platform/starboard/BUILD.gn
Comment (TyHolc) [reply ui/ozone/platform/starboard/BUILD.gn]: Done.

**Comment by TyHolc** [comment]
> New code module - should have a a unit test. ( channeling my inner @yell0wd0g )
> 
> You can't _functionally_ test a lot here, as there's no obvious way to mock EGL, but at least we should have test code calling into the function.

I added tests for surface factory, but I was blocked from testing any gl_ozone_egl_starboard due to it crashing because starboard was uninitialized. Even trying to start a starboard thread in the test didn't seem to work, and from the meeting this morning it sounds like we'll have a more concrete and compartmentalized way to initialize starboard coming soon.

File: ui/ozone/platform/starboard/BUILD.gn
Comment (TyHolc) [reply ui/ozone/platform/starboard/BUILD.gn]: Oh actually it looks like this is the cause of the new build errors:

```
ERROR at //ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc:19:11: Can't include this header from here.
#include "starboard/egl.h"
          ^--------------
The target:
  //ui/ozone/platform/starboard:starboard
is including a file from the target:
  //starboard:starboard_headers_only

It's usually best to depend directly on the destination target.
In some cases, the destination target is considered a subcomponent
of an intermediate target. In this case, the intermediate target
should depend publicly on the destination to forward the ability
to include headers.

Dependency chain (there may also be others):
  //ui/ozone/platform/starboard:starboard -->
  //starboard:starboard_group --[private]-->
  //starboard:starboard -->
  //starboard/common:common -->
  //starboard:starboard_headers_only
```

Is there something else I need to do beside swap that to starboard_group?

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc:27]: This is not 60Hz like l.24 says (display refresh rates come in Hz, not in fps).

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc:34]: `= default` ?

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.h
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/gl_ozone_egl_starboard.h:25]: `explicit` is used to disable the automatic use of one-argument constructors, since this constructor has no arguments, is not needed.

File: ui/ozone/platform/starboard/surface_factory_starboard.cc
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/surface_factory_starboard.cc:21]: With the code as is, `egl_implementation_` is initialized twice, once using a no-arguments constructor, then by l.20 (see Scott Meyer's Effective C++: 55 Specific Ways to Improve Your Programs and Designs -- "Item 12 prefer initialization to assignment in constructors"). Instead, create it in the initialization list.

Another question I have is why a `unique_ptr<>`? Could it be a normal variable and then return its address in l.37 ?

File: ui/ozone/platform/starboard/surface_factory_starboard.cc
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/surface_factory_starboard.cc:36]: l.28 doesn't list kGLImplementationEGLANGLE as an allowed implementation, so we should not handle it here IIUC.

File: ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc:23]: `= default;` ?

File: ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc:35]: Does this need to be a `unique_ptr<>` or can it be a normal member variable?

File: ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc:76]: All of these test/cases are calling GetGLOzone() with a given implementation, then expecting a nullptr or not. This should be done with a parameterized test created with TEST_P, and using a list of pairs kGLImplementation + expected output. 

See the multiple examples in Chrome, e.g.
https://source.chromium.org/chromium/chromium/src/+/main:media/gpu/svc_layers_unittest.cc;l=304-305 and https://source.chromium.org/chromium/chromium/src/+/main:media/gpu/svc_layers_unittest.cc;l=492-496 .

**Review by yell0wd0g — CHANGES_REQUESTED** [review]
Code is looking mostly fine, with minor nits and hooray for the unit tests! Now to parameterize them.

I'm not 100% sure of the need for CreateOffscreenGLSurface() implementation but if other Ozone backends do it, we probably should.

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc]: Done.

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc]: Done.

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.h
Comment (TyHolc) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.h]: Done.

File: ui/ozone/platform/starboard/surface_factory_starboard.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/surface_factory_starboard.cc]: Done, but we may need to look at supporting that for Linux development. IIRC I needed that line for the prototype to work on my machine.

File: ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc]: Updated this to initialize the class variable in the constructor instead of SetUp.

File: ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc]: Done.

**Review by yell0wd0g — CHANGES_REQUESTED** [review]
LGTM modulo questions below. Ship it !

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc:37]: Neat, I didn't know about custom deleters.

Looking at existing usage of SbWindow it seems like it is generally designed to be used without a pointer (all of the functions that have a SbWindow param are by value). Would we still want to use a unique_ptr and custom deleter in that case, or just call SbWindowDestroy it in the destructor here?

File: ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/test/surface_factory_starboard_unittest.cc]: Done.

File: ui/ozone/platform/starboard/surface_factory_starboard.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/surface_factory_starboard.cc]: Done.

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc]: For now I've implemented without unique_ptr/deleter to follow other usage I'm seeing.

File: ui/ozone/platform/starboard/surface_factory_starboard.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/surface_factory_starboard.cc]: Looping back around to this, I applied this to the prototype and it required another hacky workaround (for now, looking into getting actual solutions to the 2 existing ones and this new one) to avoid it setting the the gl implementation to ANGLE in gl_display.cc, regardless of the flags I tried. It'll probably be a few cherry picks once I can find where that's fixed.

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (yell0wd0g) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc]: I'd advocate for the unique_ptr<> because in general it's best to not have explicit destruction. That in itself justifies the extra cost of having a heap-allocated objec(as opposed to stack objects).

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (yell0wd0g) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc:75]: Sounds good, I wasn't sure myself.

**Review by yell0wd0g — CHANGES_REQUESTED** [review]
Still LGTM modulo the unique_ptr<> with custom deleter.

File: ui/ozone/platform/starboard/surface_factory_starboard.cc
Comment (yell0wd0g) [inline ui/ozone/platform/starboard/surface_factory_starboard.cc:19]: `: egl_implementation_() {}` is unnecessary since `egl_implementation_` is a fully owned object with a no-arg constructor so instead do just `= default;`. Same for the dtor in l.21.

File: ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc
Comment (TyHolc) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc]: I'm running into issues with this implementation. The `unique_ptr` expects dynamic allocation, and with `SbWindowCreate` we are dynamically allocating `SbWindowPrivate`, but not `SbWindow` (because it's treated as an object). Using `unique_ptr<SbWindowPrivate>` *would* work since we can cast SbWindow->SbWindowPrivate*, but we don't have the definition of `SbWindowPrivate` outside of Starboard so it fails due to an undefined type. I also tried `unique_ptr<void>` and learned that the compiler rejects that.

I spent some time working with @y4vor to look at this and we could get it working with a bunch of `reinterpret_cast`s every time we set or accessed it, but we decided that was not clean or preferred.

Comment (yell0wd0g) [reply ui/ozone/platform/starboard/gl_ozone_egl_starboard.cc]: I see, well, in the interest of time let's not block this CL on this specific lifetime mgmt.

**Review by yell0wd0g — APPROVED** [review]

**Comment by yell0wd0g** [comment]
> New code module - should have a a unit test. ( channeling my inner @yell0wd0g )
> 
> You can't _functionally_ test a lot here, as there's no obvious way to mock EGL, but at least we should have test code calling into the function.

@TyHolc  added surface_factory_starboard_unittest.cc after this comment. I'm not sure if Github will let this be merged with a pending comment, hence my bringing this up.

**Comment by TyHolc** [comment]
@kaidokert are we good to merge or were there additional changes (currently blocked by your requested changes)?

**Review by kaidokert — APPROVED** [review]

**Pull request merged**
